r/src/array_from_vector.cpp

```diff
@@ -0,0 +1,1014 @@
+
+  std::shared_ptr<Array> out;
+  STOP_IF_NOT_OK(DictionaryArray::FromArrays(type, array_indices, &out));
+  return out;
+}
+
+std::shared_ptr<Array> MakeFactorArray(Rcpp::IntegerVector_ factor, const std::shared_ptr<arrow::DataType>& type) {
+  SEXP levels = factor.attr("levels");
+  int n = Rf_length(levels);
+  if (n < 128) {
+    return MakeFactorArrayImpl<arrow::Int8Type>(factor, type);
+  } else if (n < 32768) {
+    return MakeFactorArrayImpl<arrow::Int16Type>(factor, type);
+  } else {
+    return MakeFactorArrayImpl<arrow::Int32Type>(factor, type);
+  }
+}
+
+template <typename T>
+int64_t time_cast(T value);
+
+template <>
+inline int64_t time_cast<int>(int value) {
+  return static_cast<int64_t>(value) * 1000;
+}
+
+template <>
+inline int64_t time_cast<double>(double value) {
+  return static_cast<int64_t>(value * 1000);
+}
+
+}
+}
+
+// ---------------- new api
+
+
+
+
+
+namespace arrow{
+using internal::checked_cast;
+
+namespace internal{
+
+template <typename T, typename Target>
+Status int_cast(T x, Target* out) {
+  if (x < std::numeric_limits<Target>::min() || x > std::numeric_limits<Target>::max()) {
+    return Status::Invalid("Value is too large to fit in C integer type");
+  }
+  *out = static_cast<Target>(x);
+  return Status::OK();
+}
+
+template <typename Int>
+Status double_cast(Int x, double* out) {
+  *out = static_cast<double>(x);
+  return Status::OK();
+}
+
+template <>
+Status double_cast<int64_t>(int64_t x, double* out) {
+  constexpr int64_t kDoubleMax = 1LL << 53;
+  constexpr int64_t kDoubleMin = -(1LL << 53);
+
+  if (x < kDoubleMin || x > kDoubleMax) {
+    return Status::Invalid("integer value ", x, " is outside of the range exactly",
+      " representable by a IEEE 754 double precision value");
+  }
+  *out = static_cast<double>(x);
+  return Status::OK();
+}
+
+// used for int and int64_t
+template <typename T>
+Status float_cast(T x, float* out) {
+  constexpr int64_t kHalfFloatMax = 1LL << 24;
+  constexpr int64_t kHalfFloatMin = -(1LL << 24);
+
+  int64_t x64 = static_cast<int64_t>(x);
+  if (x64 < kHalfFloatMin || x64 > kHalfFloatMax) {
+    return Status::Invalid("integer value ", x, " is outside of the range exactly",
+      " representable by a IEEE 754 half precision value");
+  }
+
+  *out = static_cast<float>(x);
+  return Status::OK();
+}
+
+template <>
+Status float_cast<double>(double x, float* out) {
+  //  TODO: is there some sort of floating point overflow ?
+  *out = static_cast<float>(x);
+  return Status::OK();
+}
+
+}
+
+namespace r{
+
+class VectorConverter;
+
+Status GetConverter(const std::shared_ptr<DataType>& type, std::unique_ptr<VectorConverter>* out);
+
+class VectorConverter {
+public:
+  virtual ~VectorConverter() = default;
+
+  virtual Status Init(ArrayBuilder* builder) = 0;
+
+  virtual Status Ingest(SEXP obj) = 0;
+
+  virtual Status GetResult(std::shared_ptr<arrow::Array>* result) {
+    RETURN_NOT_OK(builder_->Finish(result));
+    return Status::OK();
```

Comment:
you can `return builder_->Finish(result)`

Reply:
Thanks. Will do in a next PR.